File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/paywall/TieredPaywallCatalog.kt
package com.wagerproof.app.features.paywall

import com.wagerproof.core.models.SubscriptionTier
import java.net.URI
import java.util.Locale
import kotlin.math.roundToInt

internal object TieredPaywallCatalog {
    const val OFFERING_ID = "wagerproof_tiers_v1"
    fun priceCents(tier: SubscriptionTier, yearly: Boolean) =
        (if (yearly) listOf(19999, 29999, 35999) else listOf(1999, 2999, 7999))[tier.ordinal]
    fun comparisonCents(tier: SubscriptionTier, yearly: Boolean): Int? =
        when(tier) { SubscriptionTier.STANDARD -> null; SubscriptionTier.PREMIUM -> 5999; SubscriptionTier.PRO -> 12999 }?.times(if(yearly) 12 else 1)
    fun packageId(tier: SubscriptionTier, yearly: Boolean) = "${tier.slug}_${if(yearly) "yearly" else "monthly"}"
    fun productId(tier: SubscriptionTier, yearly: Boolean) = "wp_tiers_${tier.slug}:${if(yearly) "yearly" else "monthly"}"
    fun usd(cents: Int) = String.format(Locale.US, "$%.2f", cents / 100.0)
    fun savings(price: Long, comparison: Long) = if (price > 0 && comparison > price) ((1.0 - price.toDouble()/comparison)*100).roundToInt() else null
    fun hostedUrl(base: String?, rcUser: String, authUser: String?): String? {
        if (authUser == null || rcUser != authUser.lowercase() || runCatching { java.util.UUID.fromString(authUser) }.isFailure) return null
        val uri = runCatching { URI(base ?: return null) }.getOrNull() ?: return null
        if (uri.scheme != "https" || uri.host != "pay.rev.cat" || uri.userInfo != null || uri.port != -1 || uri.query != null || uri.fragment != null || uri.path.trim('/').contains('/') || uri.path.trim('/').isEmpty()) return null
        return "https://pay.rev.cat/${uri.path.trim('/')}/$rcUser?currency=USD"
    }
    data class Feature(val id: String, val title: String, val detail: String, val tier: SubscriptionTier)
    val features = listOf(
        Feature("games", "Game Data & Matchups", "Every matchup, key stats, injury reports, and live scores in one place.", SubscriptionTier.STANDARD),
        Feature("predictions", "WagerProof Predictions", "Model projections and signals to help you research every game.", SubscriptionTier.STANDARD),
        Feature("markets", "Prediction Markets", "Follow market probabilities and see how they compare with our models.", SubscriptionTier.STANDARD),
        Feature("discord", "Discord Community", "Join the WagerProof community to discuss games, research, and strategy.", SubscriptionTier.STANDARD),
        Feature("outliers", "Outliers", "Spot the biggest gaps between model projections and market expectations.", SubscriptionTier.PREMIUM),
        Feature("props", "Player Props", "Research player lines, matchup context, and recent performance.", SubscriptionTier.PREMIUM),
        Feature("parlays", "Parlay God", "Explore curated parlay combinations with the research behind every leg.", SubscriptionTier.PREMIUM),
        Feature("cheats", "Cheat Sheets", "Find the strongest trends and prop opportunities in a quick, scannable view.", SubscriptionTier.PREMIUM),
        Feature("agents", "Personal AI Agents", "Build up to 30 agents around your strategies, with 8 active in your office.", SubscriptionTier.PRO),
        Feature("picks", "Agent Picks & Autopilot", "Get picks from your agents and let autopilot handle their daily research.", SubscriptionTier.PRO),
        Feature("leaderboard", "Full Agent Leaderboard", "Explore every rank, compare track records, and follow the agents you trust.", SubscriptionTier.PRO),
    )
}

File: wagerproof-ios-native/Wagerproof/Features/Paywall/TieredPaywallCatalog.swift
import Foundation
import WagerproofModels

/// Opt-in experiment. Existing subscribers and the default paywall are unchanged.
enum TieredPaywallConfiguration {
    // Enables the renderer; RevenueCat placement targeting still controls enrollment.
    static let enabled = true
    static let offeringID = "wagerproof_tiers_v1"

    /// Purchase diagnostics are available in development and TestFlight only.
    static var purchaseTestingAvailable: Bool {
        #if DEBUG
        true
        #else
        Bundle.main.appStoreReceiptURL?.lastPathComponent == "sandboxReceipt"
        #endif
    }

    /// Device-local design testing in the real onboarding host. Never enables
    /// purchases or changes the production placement audience.
    static var onboardingPreviewEnabled: Bool {
        #if DEBUG
        UserDefaults.standard.bool(forKey: "tieredOnboardingPreview")
        #else
        false
        #endif
    }
}

enum PaywallTier: Int, CaseIterable, Identifiable {
    case standard, premium, pro
    var id: Int { rawValue }
    var slug: String { ["standard", "premium", "pro"][rawValue] }
    var title: String { accessTier.title }
    var summary: String {
        ["Know the game. Follow the market.", "Find your edge with deeper research.", "Your strategies. Your team of AI agents."][rawValue]
    }
    func priceCents(yearly: Bool) -> Int {
        (yearly ? [19999, 29999, 35999] : [1999, 2999, 7999])[rawValue]
    }
    func webPriceCents(yearly: Bool) -> Int {
        // Round half-up to the nearest cent, matching the web catalog.
        (priceCents(yearly: yearly) * 70 + 50) / 100
    }
    /// Yearly comparisons annualize the higher USD monthly reference price.
    /// Other storefront currencies must use their own localized comparison.
    func usdCompareAtCents(yearly: Bool) -> Int? {
        let monthly: Int
        switch self {
        case .standard: return nil
        case .premium: monthly = 5999
        case .pro: monthly = 12999
        }
        return monthly * (yearly ? 12 : 1)
    }
    func packageID(yearly: Bool) -> String { "\(slug)_\(yearly ? "yearly" : "monthly")" }
    func productID(yearly: Bool) -> String { "com.wagerproof.mobile.tiers.\(packageID(yearly: yearly))" }
    var accessTier: SubscriptionTier { SubscriptionTier(rawValue: slug)! }
    var entitlementID: String { accessTier.entitlementID }
    static func usd(_ cents: Int) -> String { String(format: "$%.2f", Double(cents) / 100) }
}

struct TieredPaywallFeature: Identifiable {
    let id: String
    let icon: String
    let title: String
    let detail: String
    let tier: PaywallTier

    static let previewFeatures: [Self] = [
        .init(id: "games", icon: "sportscourt", title: "Game Data & Matchups", detail: "Every matchup, key stats, injury reports, and live scores in one place.", tier: .standard),
        .init(id: "predictions", icon: "chart.xyaxis.line", title: "WagerProof Predictions", detail: "Model projections and signals to help you research every game.", tier: .standard),
        .init(id: "markets", icon: "chart.bar.xaxis", title: "Prediction Markets", detail: "Follow market probabilities and see how they compare with our models.", tier: .standard),
        .init(id: "discord", icon: "bubble.left.and.bubble.right", title: "Discord Community", detail: "Join the WagerProof community to discuss games, research, and strategy.", tier: .standard),
        .init(id: "outliers", icon: "bolt.circle", title: "Outliers", detail: "Spot the biggest gaps between model projections and market expectations.", tier: .premium),
        .init(id: "props", icon: "figure.basketball", title: "Player Props", detail: "Research player lines, matchup context, and recent performance.", tier: .premium),
        .init(id: "parlays", icon: "square.stack.3d.up", title: "Parlay God", detail: "Explore curated parlay combinations with the research behind every leg.", tier: .premium),
        .init(id: "cheats", icon: "list.bullet.rectangle", title: "Cheat Sheets", detail: "Find the strongest trends and prop opportunities in a quick, scannable view.", tier: .premium),
        .init(id: "agents", icon: "person.crop.square", title: "Personal AI Agents", detail: "Build up to 30 agents around your strategies, with 8 active in your office.", tier: .pro),
        .init(id: "picks", icon: "checkmark.seal", title: "Agent Picks & Autopilot", detail: "Get picks from your agents and let autopilot handle their daily research.", tier: .pro),
        .init(id: "leaderboard", icon: "trophy", title: "Full Agent Leaderboard", detail: "Explore every rank, compare track records, and follow the agents you trust.", tier: .pro),
    ]
}


/// Shared artwork mapping keeps the paywall and every feature gate consistent.
/// Existing artwork remains the Standard tier; upgrades use sibling assets.
enum TierArtwork {
    static func bolt(for tier: SubscriptionTier) -> String {
        switch tier {
        case .standard: "TieredPaywallBolt"
        case .premium: "TieredPaywallPremiumBolt"
        case .pro: "TieredPaywallProBolt"
        }
    }

    static func background(for tier: SubscriptionTier) -> String {
        switch tier {
        case .standard: "TieredPaywallArtwork"
        case .premium: "TieredPaywallPremiumArtwork"
        case .pro: "TieredPaywallProArtwork"
        }
    }
}
